### PR TITLE
#899 App Credit Rule CFA-2

### LINKED ISSUE #899: [ETHEREUM-CONTRACTS] App Credit Rule CFA-2: Minimum buffer amount as additional app credit
Per discussion in protocol forum https://www.youtube.com/watch?v=fYXjAXDRxDY.

Detailed Technical Notes: https://github.com/superfluid-finance/protocol-monorepo/wiki/About-App-Credit

### PR DESCRIPTION
**Pull request by 0xdavinchee**
* Implement new app credit rule logic
* TODO: implement new app credit rule tests
  * Fix existing tests
  * Unit tests
  * Fuzz tests
* Unify naming from allowance to credit across contracts codebase
* Minor typo fixes

### CODE REVIEW

**Comment by 0xdavinchee** [comment]
https://whimsical.com/buffer-app-credit-system-Vw2YRBPeZyaPMDuzYqMDep

**Comment by hellwolf** [comment]
Progress summary:

- ✅ the lowest bar is that there is no regression detected by existing test cases (we should not cheat neither by disabling or modifying existing ones without a reason)
- [x] One monkey testing session using MFATestApp to proof read the numbers visually.

File: packages/ethereum-contracts/utils/mfa-tester.html [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+    <title>Multi Flow SuperApp Tester</title>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
+    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.3.1/dist/css/bootstrap.min.css"
+        integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous" />
+    <script src="https://cdn.jsdelivr.net/npm/ethers@5.6.0/dist/ethers.umd.min.js"
+        integrity="sha256-I/T7C/qM1p/EIawMnVp+b68aGIgHpiaW5uJm8Xd/WYg=" crossorigin="anonymous"></script>
+
+    <!-- sdk-core -->
+    <script src="./index.umd.js"></script>
```
Comment (hellwolf) [inline packages/ethereum-contracts/utils/mfa-tester.html]: maybe create a new level of subfolder "dist", which is automatically git-ignored

File: packages/ethereum-contracts/.eslintrc.json [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
-    "ignorePatterns": ["coverage", "coverage.json", "build"]
+    "ignorePatterns": ["coverage", "coverage.json", "build", "utils"]
```
Comment (hellwolf) [inline packages/ethereum-contracts/.eslintrc.json:2]: mkay

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol [hunk 3]
```diff
@@ -480,16 +478,16 @@ interface ISuperfluid {
...
-        // app allowance granted
-        uint256 appAllowanceGranted;
-        // app allowance wanted by the app callback
-        uint256 appAllowanceWanted;
-        // app allowance used, allowing negative values over a callback session
-        int256 appAllowanceUsed;
+        // app credit granted
+        uint256 appCreditGranted;
+        // app credit wanted by the app callback
+        uint256 appCreditWanted;
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol]: mark it as appCreditWantedDeprecated and remove the usage of it throughout.

Comment (hellwolf) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol]: make some additional comments re decodeCtx that it has abi compatibility requirement.

File: packages/ethereum-contracts/contracts/agreements/AgreementLibrary.sol [hunk 4]
```diff
@@ -140,17 +140,35 @@ library AgreementLibrary {
...
+     */
+    function _adjustNewAppCreditUsed(
+        uint256 appCreditGranted,
+        int256 appCreditUsed
+    ) internal pure returns (int256) {
+        return max(
+            0,
+            min(
+                appCreditGranted.toInt256(),
+                appCreditUsed
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/AgreementLibrary.sol]: 1. add comment: appCreditUsed can be minus, hence `max(0, -)` needed
2. possible other names for appCreditUsed: app[CallbackSession]DepositDelta, comment in ISuperfluid interface about what does it mean. Make some more elaboration of its use per discussion.

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 5]
```diff
@@ -959,14 +957,33 @@ contract ConstantFlowAgreementV1 is
...
+
+            (,cbStates.appCreditGranted,) = _changeFlow(
                     currentContext.timestamp,
-                    currentContext.appAllowanceToken,
+                    currentContext.appCreditToken,
                     token, flowParams, oldFlowData);
-            cbStates.appAllowanceGranted = cbStates.appAllowanceGranted * uint256(currentContext.appCallbackLevel + 1);
-            cbStates.appAllowanceUsed = oldFlowData.owedDeposit.toInt256();
+
+            // NOTE: we do not provide additionalAppCreditAmount when cbStates.appCreditGranted is 0
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:974]: ⚠️ new logic

File: packages/ethereum-contracts/contracts/agreements/AgreementLibrary.sol (hunk 4 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/agreements/AgreementLibrary.sol]: ⚠️ new logic

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 6]
```diff
@@ -983,38 +1000,31 @@ contract ConstantFlowAgreementV1 is
...
         {
-            // clipping the allowance used amount before storing
-            if (vars.appContext.appAllowanceUsed > 0) {
-                // give more to the app
-                vars.appContext.appAllowanceUsed =
-                    _clipDepositNumber(vars.appContext.appAllowanceUsed.toUint256()).toInt256();
-            }
-
-            int256 appAllowanceDelta = vars.appContext.appAllowanceUsed
+            int256 appCreditDelta = vars.appContext.appCreditUsed
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:1013]: ⚠️ deleted logic

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 7]
```diff
@@ -1055,8 +1064,8 @@ contract ConstantFlowAgreementV1 is
                         int256 userDamageAmount = AgreementLibrary.min(
                             // user will take the damage if the app is broke,
                             -availableBalance,
-                            // but user's damage is limited to the amount of app allowance it gives to the app
-                            AgreementLibrary.max(0, -appAllowanceDelta));
+                            // but user's damage is limited to the amount of app credit it gives to the app
+                            -appCreditDelta);
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:1071]: ⚠️ new logic. supposedly appCreditDelta is always positive

File: packages/ethereum-contracts/hardhat.config.ts [hunk 8]
```diff
@@ -49,7 +49,7 @@ const chainIds = {
 
     "bsc-mainnet": 56,
 
-    localhost: 1337,
+    localhost: 31337,
```
Comment (hellwolf) [inline packages/ethereum-contracts/hardhat.config.ts:52]: 👟

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js [hunk 9]
```diff
@@ -3239,9 +3237,19 @@ describe("Using ConstantFlowAgreement v1", function () {
             await timeTravelOnceAndVerifyAll();
         });
 
-        it("#2.6 mfa-1to1-101pct_create-should-fail-without-extra-funds", async () => {
+        it("#2.6 mfa-1to1-?pct_create-should-fail-without-extra-funds", async () => {
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js:3240]: "?" ??

Comment (hellwolf) [reply packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js:3240]: ⚠️ review test case

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js [hunk 10]
```diff
@@ -3703,10 +3711,13 @@ describe("Using ConstantFlowAgreement v1", function () {
                 "CFA: sender account is not critical"
             );
 
+            // @note - need to consider new additional app credit rule CFA-2
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js:3737]: ⚠️ review test case

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js (hunk 9 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js:3240]: added more comments to explain changes

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js (hunk 10 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js:3737]: added more comments to explain changes

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol (hunk 7 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:1071]: added comments to explain why this should be the case

**Review by hellwolf — APPROVED** [review]
let's gooo

**Comment by hellwolf** [comment]
need to work on technical notes

**Pull request merged**